apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+import {* as stepper} from 'stepper';
+
+const store = createStore(
+  combineReducers({stepperState: stepper.reducer})
+);
+
+// tell the stepper how to access its portion of the global state tree.
+stepper.configure({getState: state => state.stepperState});
+
+const StepBanner = connect(
+  state => ({step: stepper.selectors.getStep(state)})
+)(props => <div>You are on step {props.step}</div>);
+
+ReactDOM.render(
+  <Provider store={store}>
+    <StepBanner />
+  </Provider>
+);
+```
+
+#### Exposing/Providing Connected React Components ####
+
+Features that maintain shared application state usually do so because they also
+render/expose connected react components that work with that state. Continuing
+with our stepper library example, if we wanted to export a connected
+`<Incrementor />` component, we would write it like this:
+
+**`src/stepper/components/Incrementor.jsx`**
+```jsx
+import {incrementStep} from '../redux/actionCreators';
+import {getStep} from '../redux/selectors';
+
+connect(
+  state => ({step: getStep(state)}),
+  {incrementStep}
+)(props => <button onClick={props.incrementStep}>Increment to {props.step + 1}</button>);
+```
+
+and the `index.js` file would be updated to explicitly export this component:
+
+**`src/stepper/index.js`**
+
+```js
+import Incrementor from './components/Incrementor';
+export Incrementor;
+```
+
+#### Using Shared State in a Feature Library ####
+
+In some circumstances, feature libraries may need to access shared state that is
+provided by another feature library. To make our discussion more concrete, let's
+assume that our stepper feature library needs to check if a user is logged in
+before allowing the step to be incremented. The login state of the application
+is provided by an authentication feature library.
+
+If the stepper library imports selectors directly from the authentication
+library to access the logged in state, then there is an implicit dependency
+created between the stepper library and the entry point that uses it because it
+assumes that at some point the authentication library has been configured
+properly by the entry point. This would be unintentional coupling.
+
+Instead, the entry point should be responsible for passing the configured
+authentication selectors directly to the stepper library via it's `configure`
+function. Here is how that might look:
+
+**`src/entry-points/some-page.js`**
+Now our entry point has to configure both the `stepper` and the `auth`
+libraries and manage how they access each other's shared state.
+
+```js
+import {createStore, combineReducers} from 'redux';
+import {* as auth} from '../auth';
+import {* as stepper} from '../stepper';
+
+const store = createStore(
+  combineReducers({
+    authState: auth.reducer,
+    stepperState: stepper.reducer
+  })
+);
+
+auth.configure({
+  getState: state => state.authState,
+});
+stepper.configure({
+  getState: state => state.stepperState,
+  authSelectors: auth.selectors,
```

Comment:
I would even be comfortable with injecting all of `auth` here, since you're still bound to the interface `auth` exposes.  Doubly so for libraries where we instantiate something instead of creating a singleton.